### PR TITLE
RavenDB-12520 _ensure_path_exists implemented for posix

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-12520

### CODE REVIEW

File: src/librvnpal/src/posix/fileutils.c [hunk 1]
```diff
@@ -86,10 +87,67 @@ _allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code)
...
+            *detailed_error_code = ENOTDIR;
+            rc = FAIL_NOT_DIRECTORY;
+            goto error_cleanup_with_error;
+        }
+
+        if (current_end != NULL)
+            *current_end = '/';
+    }
+
+    return SUCCESS;
```
Comment (ayende) [inline src/librvnpal/src/posix/fileutils.c]:
```suggestion
    rc = SUCCESS;
    goto error_cleanup_with_error;
```

Here you are leaking `dup_path` on success.

File: src/librvnpal/src/posix/mapping.c [hunk 2]
```diff
@@ -50,15 +51,20 @@ rvn_create_and_mmap64_file(
...
-    struct map_file_handle* mfh = calloc(1, sizeof(struct map_file_handle));
+    struct map_file_handle *mfh = calloc(1, sizeof(struct map_file_handle));
     *handle = mfh;
     if(mfh == NULL)
     {
         rc = FAIL_CALLOC;
         goto error_cleanup;
     }
+
+    char* directory = strdup(path);
```
Comment (ayende) [inline src/librvnpal/src/posix/mapping.c]: Need to check if we failed here

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/librvnpal/src/posix/fileutils.c (hunk 1 above)
Comment (aviviadi) [inline src/librvnpal/src/posix/fileutils.c]: free dup_path missing

File: src/librvnpal/src/posix/fileutils.c [hunk 3]
```diff
@@ -86,10 +87,67 @@ _allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code)
...
 {
-    /* TODO: implement */
-    return SUCCESS;   
+    assert(path != NULL);
+
+    int32_t rc;
+    char* dup_path = strdup(path);
+    if(dup_path == NULL)
+    {
+        *detailed_error_code = errno;
```
Comment (aviviadi) [inline src/librvnpal/src/posix/fileutils.c]: goto errorCleanup

File: src/librvnpal/src/posix/fileutils.c [hunk 4]
```diff
@@ -86,10 +87,67 @@ _allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code)
...
+        if (current_end != NULL)
+            *current_end = '/';
+    }
+
+    return SUCCESS;
+
+error_cleanup:
+    *detailed_error_code = errno;
+error_cleanup_with_error:
+    free(dup_path);
```
Comment (aviviadi) [inline src/librvnpal/src/posix/fileutils.c]: verify dup_path != NULL

File: src/librvnpal/src/posix/fileutils.c (hunk 3 above)
Comment (ayende) [reply src/librvnpal/src/posix/fileutils.c]: Don't think that would be a good idea. It calls to `free(dup_path)`, and that is NULL.
Not sure what will happen, but not valid option to do

File: src/librvnpal/src/posix/fileutils.c [hunk 5]
```diff
@@ -86,10 +87,67 @@ _allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code)
...
+        
+        if (stat(dup_path, &sb) == -1)
+        {
+            if (errno != ENOENT)
+            {
+                rc = FAIL_STAT_FILE;
+                goto error_cleanup;
+            }
+
+            if (mkdir(dup_path, 0700) == -1)
```
Comment (aviviadi) [inline src/librvnpal/src/posix/fileutils.c]: use 0755 permissions (like dotnet's System.IO.Directory.CreateDirectory)

File: src/librvnpal/src/posix/mapping.c [hunk 6]
```diff
@@ -50,15 +51,20 @@ rvn_create_and_mmap64_file(
...
+    int ret = _ensure_path_exists(dirname(directory), detailed_error_code);
```
Comment (aviviadi) [inline src/librvnpal/src/posix/mapping.c]: 1. verify path != null (however we do allow empty string)
2. dirname will return "." - verify _ensure_path_exists can handle string "."

**Comment by ayende** [comment]
Also, write some tests here for adversarial input. Can we pass non path arguments here somehow? How does it handle it?

File: src/librvnpal/src/posix/mapping.c [hunk 7]
```diff
@@ -48,17 +49,32 @@ rvn_create_and_mmap64_file(
...
-    struct map_file_handle* mfh = calloc(1, sizeof(struct map_file_handle));
+    struct map_file_handle *mfh = calloc(1, sizeof(struct map_file_handle));
     *handle = mfh;
     if(mfh == NULL)
     {
         rc = FAIL_CALLOC;
         goto error_cleanup;
     }
+
+    if(path[0] == '\0')
```
Comment (ayende) [inline src/librvnpal/src/posix/mapping.c]: How can that be? What would be passing us something like this?
And if they are passing us an empty string, wouldn't this cause issues with `strdup` never ending?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/librvnpal/src/posix/mapping.c (hunk 7 above)
Comment (haludi) [reply src/librvnpal/src/posix/mapping.c]: https://github.com/ravendb/ravendb/pull/8970#discussion_r266767782
@avivadi commented that we should accept an empty string and it should reference to current directory as `"."`. As I understand it `strdup` has no problem with an empty string but `stat` do
so I switch it with `"."`.

Comment (ayende) [reply src/librvnpal/src/posix/mapping.c]: Consider the case where we send an empty C# string `""`.
This is translated by pinvoke to `"\0"` as a C string.
You are writing changing this to be: `"."`, but you _don't have a null terminator.

Now you pass a string without null terminator to `strdup` and will cause _issues_.

You can _change_ the value, so `if (path[0] == '\0') path = "."; `, which _is_ null terminated.

Comment (ayende) [reply src/librvnpal/src/posix/mapping.c]: Also add a comment explaining this explicitly

**Review by ayende — APPROVED** [review]

**Comment by ppekrol** [comment]
@haludi can you check failing tests?

**Pull request closed without merging**